File: DSA/LeetCode/Arrays/448 - Find All Numbers Disappeared in an Array/Solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDisappearedNumbers = function (nums) {
  const arr = Array(nums.length).fill().map((v, i) => ++i);

  for (let num of nums) {
    arr[num - 1] = -1
  }

  return arr.filter((num) => num !== -1);
};

// Alternate solution
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDisappearedNumbers = function (nums) {
  for (let i in nums) {
    const index = Math.abs(nums[i]) - 1;
    if (nums[index] > 0) {
      nums[index] *= -1
    }
  }

  const result = []
  for (let j in nums) {
    if (nums[j] > 0) {
      result.push(++j);
    }
  }
  return result;
}

File: DSA/LeetCode/Arrays/448 - Find All Numbers Disappeared in an Array/Solution.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        arr = list(range(1, len(nums)+1))

        for index, num in enumerate(nums):
            arr[num-1] = -1
        return filter(lambda x: x != -1, arr)

# Alternate solution


class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        for i in range(len(nums)):
            index = abs(nums[i]) - 1
            if nums[index] > 0:
                nums[index] *= -1

        return [i+1 for i, n in enumerate(nums) if n > 0]
